### PR TITLE
Democracy Now news scraper ETL

### PR DESCRIPTION
**Pull request by YakMan101**
Created the ETL scripts to 
- scrape all the Democracy Now topics
- find all articles from each topic 
- extract information about all articles
- store results as dataframe
- upload dataframe as CSV to s3 bucket.

Also created an initial readme and docker file

### CODE REVIEW

File: democracy_now_scraper/extract_dn.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+def fetch_response_html(response) -> BeautifulSoup:
+    """Return html content of response"""
+
+    return BeautifulSoup(response.content, "html.parser")
+
+
+def chunk_links(links: list[str], chunk_size: int) -> list[list[str]]:
+    """
+    Split the list of links into smaller chunks.
+    """
```
Comment (indiah444) [inline democracy_now_scraper/extract_dn.py:18]: Small thing, but might want to standardise the formatting of the one line docstrings.

File: democracy_now_scraper/pipeline_dn.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            "statusCode": 200,
+            "body": "data extracted, processed and uploaded to S3 successfully."
+        }
+
+    except Exception as e:
+        return {
+            "statusCode": 500,
+            "body": f"Error: {str(e)}"
+        }
+
```
Comment (indiah444) [inline democracy_now_scraper/pipeline_dn.py:40]: Good error catching with appropriate status codes.

File: democracy_now_scraper/extract_dn.py (hunk 1 above)
Comment (YakMan101) [reply democracy_now_scraper/extract_dn.py]: Good spot

File: democracy_now_scraper/pipeline_dn.py (hunk 2 above)
Comment (YakMan101) [reply democracy_now_scraper/pipeline_dn.py:40]: I took inspiration from your pipeline script!

File: democracy_now_scraper/pipeline_dn.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            print("Could not convert results to dataframe")
+            return
+
+        current_time = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+        s3_filename = f"{current_time}_democracy_now_article_data.csv"
+        upload_dataframe_to_s3(df, s3_filename)
+
+        return {
+            "statusCode": 200,
+            "body": "data extracted, processed and uploaded to S3 successfully."
```
Comment (ayahusseini) [inline democracy_now_scraper/pipeline_dn.py:32]: Small comment but capitalise "Data"

File: democracy_now_scraper/pipeline_dn.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from datetime import datetime
+
+from dotenv import load_dotenv
+
+from extract_dn import scrape_democracy_now
+from transform_dn import convert_to_dataframe
+from load_dn import upload_dataframe_to_s3
+
+
+def lambda_handler(event, context):
```
Comment (ayahusseini) [inline democracy_now_scraper/pipeline_dn.py:10]: can we add typehints for event and context (dict)

File: democracy_now_scraper/load_dn.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    df.to_csv(csv_buffer, index=False)
+
+    s3_bucket = ENV['S3_BUCKET_NAME']
+    s3_client = boto3.client(service_name="s3",
+                             aws_access_key_id=ENV["AWS_ACCESS_KEY_BOUDICCA"],
+                             aws_secret_access_key=ENV["AWS_ACCESS_SECRET_KEY_BOUDICCA"])
+    s3_client.put_object(Bucket=s3_bucket, Key=object_name,
+                         Body=csv_buffer.getvalue())
+
+    print(f"Uploaded DataFrame to s3://{s3_bucket}/{object_name}")
```
Comment (ayahusseini) [inline democracy_now_scraper/load_dn.py:21]: Good use of logging/printing

File: democracy_now_scraper/extract_dn.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        else:
+            article_info = scrape_article(response)
+            if isinstance(article_info, str):
+                print(article_info)
+            else:
+                articles.append(article_info)
+    return articles
+
+
+def link_is_old(link: str, time_diff: int) -> bool:
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:230]: Good use of a validation function

File: democracy_now_scraper/extract_dn.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    """
+    Filter out link that are more than 1 week old.
+    Expects links in the format 'https://www.democracynow.org/YYYY/M/D/...'.
+    """
+
+    one_week_ago = datetime.now() - timedelta(days=time_diff)
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:236]: good use of descriptive variable names: the code is very readable

File: democracy_now_scraper/extract_dn.py [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+    try:
+        date_str = '/'.join(link.split('/')[3:6])
+        link_date = datetime.strptime(date_str, "%Y/%m/%d")
+        if link_date >= one_week_ago:
+            return False
+
+    except ValueError:
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:244]: Bit unsure about why you're handling ValueErrors this way?

File: democracy_now_scraper/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
+# Democracy Now News Scraper
+
+## Overview
```
Comment (megansam5) [inline democracy_now_scraper/README.md:3]: Nice readme!

File: democracy_now_scraper/extract_dn.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+from bs4 import BeautifulSoup
+
+
+def fetch_response_html(response) -> BeautifulSoup:
+    """Return html content of response"""
+
+    return BeautifulSoup(response.content, "html.parser")
+
+
+def chunk_links(links: list[str], chunk_size: int) -> list[list[str]]:
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:15]: Good use of chunking

File: democracy_now_scraper/dockerise.sh [hunk 11]
```diff
@@ -0,0 +1,7 @@
+source .env
```
Comment (megansam5) [inline democracy_now_scraper/dockerise.sh:1]: Really good idea to dockerise this!

File: democracy_now_scraper/extract_dn.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+def reformat_date(date: str) -> str:
+    """
+    Change from American date to standard YYYY-MM-DD.
+    Expecting data in "October 07, 2024" or "Oct 07, 2024".
+    """
+
+    try:
+        date_obj = datetime.strptime(date, "%B %d, %Y")
+    except ValueError:
```
Comment (megansam5) [inline democracy_now_scraper/extract_dn.py:97]: Nice catching exceptions!

File: democracy_now_scraper/extract_dn.py (hunk 8 above)
Comment (ayahusseini) [reply democracy_now_scraper/extract_dn.py:244]: - got clarification: it's to handle archived articles (which don't have a date)

**Review by megansam5 — APPROVED** [review]

File: democracy_now_scraper/extract_dn.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+def get_article_date(soup: BeautifulSoup) -> str:
+    """Return date of article given the web body"""
+
+    try:
+        article_tag = soup.find(
+            "div", class_="container-fluid", id="story_content")
+        date = article_tag.find("span", class_="date")
+        return reformat_date(date.get_text(strip=True))
+
+    except AttributeError:
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:115]: good use of error handling

File: democracy_now_scraper/extract_dn.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+    """
+    Change from American date to standard YYYY-MM-DD.
+    Expecting data in "October 07, 2024" or "Oct 07, 2024".
+    """
+
+    try:
+        date_obj = datetime.strptime(date, "%B %d, %Y")
+    except ValueError:
+        try:
+            date_obj = datetime.strptime(date, "%b %d, %Y")
```
Comment (ayahusseini) [inline democracy_now_scraper/extract_dn.py:99]: good way of trying alternative date formats

**Review by ayahusseini — CHANGES_REQUESTED** [review]
Small changes: Capitalising Error messages and adding type hints to some functions

**Comment by ayahusseini** [comment]
Looks good - thanks for making the changes!

**Review by ayahusseini — APPROVED** [review]

**Pull request merged**